src/Nest/Document/Multiple/Bulk/BulkOperation/BulkOperationsCollection.cs

```diff
@@ -0,0 +1,222 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+
+namespace Nest
+{
+	/// <summary>
+	/// This class is used by <see cref="IBulkRequest.Operations" /> which needs thread safe adding <see cref="ICollection{T}.Add" /> as well as expose
+	/// an equivalent of <see cref="List{T}.AddRange"/>. Because operations from Elasticsearch are executed in order none of the types in
+	/// System.Collection.Concurrent can't be used for this. We need to preserve insert order and exposed indexed index because <see cref="IBulkResponse.Items"/>
+	/// is ordered and lines up with <see cref="IBulkRequest.Operations"/> allowing one to zip the two together.
+	/// </summary>
+	/// <typeparam name="TOperation"></typeparam>
+	[ComVisible(false)]
+	public sealed class BulkOperationsCollection<TOperation> : IList<TOperation>, IList
+		where TOperation : IBulkOperation
+	{
+		private readonly object _lock = new object();
```

Comment:
`ReaderWriterLockSlim` ?

Reply:
Discussed this, assumption is that this collection is write heavy and will not benefit from `ReadWriterLockSlim`